### DISCUSSION TITLE
Project Update - November  2022

### DISCUSSION THREAD

**Discussion by DeeEmm** [DIY-Flow-Bench/DIY-Flow-Bench#107] [Announcements]
It's been a few months since that last update so I thought it was about time to post another.

Progress since the last update has been a bit off-and-on-again with life inconveniently getting in the way. I have made a bunch of progress but it has been fairly slow going.

## I2C Issues

The previously mentioned I2C issues have been difficult to resolve. The main issue I was seeing is that whilst I2C is supposed to support multiple slaves (up to 128) each I2C device used its own library which did not play nice with the other I2C device libraries as they mostly each include similar but conflicting I2C protocol code. The net result was bus conflicts which resulted in data reliability and stability issues. It might actually be unfair to apportion all of the blame directly on the libraries as I believe that at least part of the issue is the interrupt driven nature of the web-socket handler interrupting the data transmissions.

## New I2C Libraries

Initially I tried to resolve the library issues by rewriting the library files, but after writing three seperate libraries I ended up down the rabbit hole chasing my tail. Next I tried to implement device drivers from a sole supplier thinking that they may play nicer together, but this suffered a similar issue. I then came across the I2Cdev library by Jeff Rowberg which abstracts the I2C communication into portable class functions that can then be reused by multiple device sub classes. This seemed like a really good fit and more or less what I needed to achieve to overcome the issues I was facing, however there was no BME280 or ADS1115 device driver classes for the ESP32 platform so I needed to port some existing but functionally similar device classes from his repo. 

## Dual Core

I also experimented with moving part of the code onto core2 of the ESP (the ESP is actually a 2 core device), but this is not without some caveats - specifically that the two cores do not run completely independently of each other. Whilst there can be two distinct threads (tasks), control is passed back and forth between the threads which effectively means that only one core takes focus at a time. This does go some way towards a solution as it can be used to segregate the sensor functions from the webserver functions but still does not directly address the interrupt issue. The other caveat is that the system Wifi handler resides on core 2.

Eventually I ended up getting stuck at some runtime errors that did not generate a valuable error output, which made them impossible to diagnose. I decided that the way to get over this hurdle was to invest in a proper development board and move to an IDE that handles debugging. Something that I have long been meaning to do.

## Debugger

For a long time I have used Nova which I integrated Arduino-CLI into to take care of compilation duties (it's essentially a glorified mac based programmers text-editor). This has worked fine and a lot faster than the Arduino IDE which is a boon for dev work. However this does not natively support debugging for the ESP32. So just this past few weeks I've moved development over to VSCode / PlatformIO which supports ESP32 out of the box. 

To enable debugging you need to use a second serial connection connected to the JTAG pins (GPIO 12/13/14/15), this allows the debugger to communicate with the board at runtime and pass program and breakpoint data back and forth. A JTAG serial adaptor is generally needed for the second connection, however the official Expressif ESP-WROVER-KIT development boards actually use a different type of USB chip which has both the RPI and JTAG serial connections on the same chip which share a single USB connection to the host computer. This negates the need for jumpering a stand alone JTAG interface to the board and overcomes common connection issues.

I've spent the past few days getting to grips with the new board and trying to get the debugger working. Documentation is limited, so unfortunately it's been a bit of an uphill struggle to troubleshoot as there are only one or two experts out there in internetland that seemed to have any idea of how to set it up. So I've had to struggle along with limited info until I figured it out. Now I just need to figure out how to make the best use of it, as there's a fairly steep learning curve there too.

## Jumpers for Goalposts

Now that I have debugging working, next up is to jumper the shield to the dev board as it does not share the same Uno style footprint as the D1. I'll need to go back over the PCB schematics to figure that out. (it's been a while)

![PXL_20221112_033342024](https://user-images.githubusercontent.com/3038710/201454542-d84a772a-4ff1-4e36-876d-a7724963ea35.jpg)

Once I've got the shield hooked up I should hopefully be able to make some progress on debugging the I2C / BME / ADS issues I was previously facing. Plan is to continue with the development of I2Cdev style device libraries but to add in some semaphore (handshaking) between the web-socket interrupt routine and the main I2Cdev class to overcome any data transmission interruptions. There's a little bit of a challenge there to ensure that web-socket connections are maintained and do not drop.

## A Formal Apology

I apologise to all of those who are waiting with anticipation for a new release, but there's really no point in releasing code until it is working / stable. And whilst it may not seem as if anything has happened, there's at least a few hundred hours of dev-work since the last update. I've also been optimising and refactoring the code as I've been going along, trying to make it more efficient and easier to read.

## To Conclude, 

I am getting much closer to the finish line, but there are still some hurdles left to overcome.

Support for the Arduino IDE will still be maintained, but support for VSCode / PlatformIO will also be provided. VSCode is far superior and quite possibly an easier option for many, especially those who have no affinity for the Arduino IDE . It also supports GULP which means that it's easier to build the web interface from the raw source files than previously. At some point I will update the documentation with a walk through of setting up the project for VSCode.

I'm not too sure on timelines at this stage. I still have to get to grips with the debugger and then assuming it allows me to overcome the roadblock I was facing with the I2C issues there are still two device libraries to finish off. If I work on it full time that's at least two weeks away, but the reality is that it will be much longer than that. 

Whatever the progress I think that I need to provide more regular updates and so I would like to commit to a monthly update. I also want to get the current code uploaded to the dev branch. Whilst it's not what I would classify as 'working', hopefully there's an opportunity to enlist some help. As always, if anyone wants to contribute, please get in touch. It's a hard slog developing this solo. 

/DM

**Comment by DeeEmm** [comment]
Just to add. I've now updated the Wiki with instructions on how to set up VScode / PlatformIO. In all honesty when you read the two procedures side by side VSCode seems a lot easier, even for the novice. 

I've also committed the current code to the Dev Branch. It will compile and run but the I2C stuff is a work in progress and currently disabled. If you are interested in what's changed, checkout the changelog and take a look.

The same code is also on the WIP branch. My plan here is to test the removal of websocket support and replace it with 'Event Source' server-sent events. This is a one-way protocol which only allows data to be pushed to the browser it cannot listen for incoming requests. What this will do is remove the webSocket listen interrupt from the server code which I suspect is what is interrupting the I2C data reads.

By removing the websockets we do lose the ability for the browser to push data back to the server. This impacts things like the on/off buttons and saving the configuration data, however there are more traditional methods for achieving the same without needing to use websockets.

There's a little work to get this implemented, I'll try and get it knocked off over the coming week.

/DM

**Comment by DeeEmm** [comment]
Websockets are now gone in favour of Server-Side-Events (SSE) this has vastly simplified server to browser comms. I have also addressed the issues with the on/off buttons and aded a few additional API calls via a simple URL based API schema.

The URL API schema is as follows

```
/api/bench/on
/api/bench/off
/api/debug/on
/api/debug/off
/api/dev/on
/api/dev/off
/api/reboot
```

Each request replies with a simple JSON response that can be used as confirmation. Creating this simple schema has allowed the browser client to control server side code.

Now that the basic underling code for the move away from websockets has been done. Next up is the I2C code. 

The latest changes are on the WIP branch

**Comment by DeeEmm** [comment]
Status update:

https://github.com/DeeEmm/DIY-Flow-Bench/commit/9c63ea7b318d6a70a9dec25bf4810f55fa093ad1

Project compiles and runs but I2C is still returning unstable data despite isolating I2C to just BME
The tinyBME library returns valid stable data when run in isolation from rest of code
Data becomes corrupt when run along side ESPAsyncWebserver
Interwebs has lots of comments RE AsyncWS + I2C corruption, so this could be one potential cause.
Could also still be related to synchronous tasks / interrupts. I am unsure what ESPAsyncWS does under the hood.
Need to spend some time working through disabling each part of the code until I can isolate the cause.
Then I can engineer a fix.

I have written several I2C libraries to try and overcome the issue, but perhaps what I really need to do
is to write new webserver code. 

/DM

**Comment by DeeEmm** [comment]
Status Update:

Commit https://github.com/DeeEmm/DIY-Flow-Bench/commit/be52c80f607eb7b14e212b6d40aa2a9dee43e495 addresses I2C issue for BME. Now getting stable valid readings using the tinyBME library.

The issue was a bit obtuse. It's actually related to variable / class scope. Here's what I think was happening...

The BME class instance was being created locally within the sensor read method, which means that every time there was a sensor read, the read method was called, the class instance was created, the reading was taken and then the class instance was automatically destroyed when the result was returned from the method. 

This would normally be fine, however there is a known issue with the BME sensors relating to its read buffers in that the very first read returns the buffer value, rather than what the sensor currently reads. So the very first read could return anything depending on what currently resides in the buffer.

I moved the class instance creation out into the global space so that it persists across all sensor read methods and the issue is now solved. ( W00T!! ).

I've been scratching my head on this one for the past week so it's nice to be able to move on.

Next up is to breadboard the ADS1115 ADC and repeat the same tests but with both I2C devices in circuit. I reduced testing back to basics by removing the shield and connecting the BME directly to eliminate potential hardware issues contributing to the problem I was seeing. I will follow the same path with the ADS1115 and when I am happy that the readings are stable I will repeat the same tests with the shield.

![PXL_20221124_023300224 MP](https://user-images.githubusercontent.com/3038710/203682059-1b153fcb-eaec-427f-9fd9-ad311bd8dd1d.jpg)

A note on the ESP-ROVER-KIT board. It is not as stable as I had hoped. Running it in debug mode is fairly slow and not without its own fair share of issues, most of which relate to breakpoints. Often it is quicker to debug the old fashioned way with serialPrint commands rather than have to debug the debugger. Kinda ironic really lol. 

Regardless, I'm happy that the issue is now hopefully resolved.

/DM

**Comment by DeeEmm** [comment]
Status Update:

Commit https://github.com/DeeEmm/DIY-Flow-Bench/commit/999ca52f8d910502190c5011c27ff35dd9644a27 addresses the ADC / I2C issue. Pressure sensor inputs are working on the breadboard. These still need to have data validation when the shield is reconnected. 

Currently working through the MAF decode.

/DM

**Comment by dobrowolskibartosz** [comment]
Wow, great to see updates. Since I'm not that techy ( software side) please tell me if this is the right moment to start preparing hardware wise - or is is still worth waiting. Do you plan to update the pcb?

**Comment by DeeEmm** [comment]
Yes, it's long overdue :D 

The basic hardware schema should not change but there is a planned update to the existing shield PCB files to address some issues that have been picked up in testing. I'm holding off releasing the update until I've finalised the testing of the new code with the revised shield as there may be more undiscovered issues to address.

If you are breadboarding your own solution then it's probably easy enough to incorporate the changes as you go and if you want to open up another discussion I can walk you through those, but if you want to use the shield then I would hold off until the updated version is published.

/DM

**Reply by dobrowolskibartosz** [reply]
Great - i plan to use your shield so I will wait for the new pcb design, will it still base on the same ESP32 board?

**Reply by DeeEmm** [reply]
Yes any ESP32 style board with an Uno footprint should work.  The smaller footprint ESP32 boards generally do not have enough exposed pins so the UNO footprint works best, plus the physical size of the shield is largely determined by the pressure sensors and other components so it makes no sense to use the smaller ESP32's as the larger UNO footprint easily fits.

I've been using the Wemos D1 so far but I also have an Arducam ESP32S UNO to test with, which is basically the same footprint but has integrated SD card support. I've been using the Arducam board on my CNC's for the additional memory that the SD card gives. It's a nice option in that regards as the SD card reader is mounted directly to the board.

There is a plan on the roadmap to eventually add SD card support for data logging as the on-board SPIFFS memory has limited space, so I plan to create a pin map for the Arducam board before the next release even if the SD portion of the code does not get written until much later.

/DM

**Comment by DeeEmm** [comment]
Status Update:

Commit https://github.com/DeeEmm/DIY-Flow-Bench/commit/ce2508fcd802b8d4b7f6ee5bf912a418d6a3b746 Concludes the major code rewrite and now has functional MAF code. The code is returning valid in-range values but ideally needs to be tested with a calibrated reference to determine accuracy.

At this stage only one MAF Data file is working (default - ACDELCO_92281162.cpp). All other files need to be migrated to the new format. I will do this before the official release. One other caveat is that only voltage based MAF sensors are supported at this stage. This will not be implemented before the release (mostly as I do not have a sensor to test with)

There's still some tidying up to do with the code but a new RC release is now within sight.

Yay!

/DM

**Comment by DeeEmm** [comment]
December update: https://github.com/DeeEmm/DIY-Flow-Bench/discussions/109
